File: 0011-container-with-most-water/0011-container-with-most-water.go
/*
Time - O(n)
Space - O(1)
*/
func maxArea(height []int) int {
    left:=0
    right:=len(height)-1
    maxArea:=0
    for left<right {
        area:=(right-left)*min(height[left], height[right])
        if area>maxArea {
            maxArea=area
        }
        if height[left]<height[right] {
            left++
        }else {
            right--
        }
    }
    return maxArea
}

func min(i,j int) int {
    if i<j {
        return i
    }
    return j
}

File: 0011-container-with-most-water/0011-container-with-most-water.py
class Solution:
    def maxArea(self, height: List[int]) -> int:
        l=0
        r=len(height)-1
        maxArea=0
        while l<r:
            if height[l]<height[r]:
                maxArea=max(maxArea, height[l]*(r-l))
                l+=1
            else:
                maxArea=max(maxArea, height[r]*(r-l))
                r-=1
        return maxArea
            